### PR TITLE
HW04 Gerasev Georgii

### PR DESCRIPTION
**Pull request by VenInf**

### CODE REVIEW

File: Main.hs [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+newtype MyArrow a b = MyArrow((->) a b)
+
+instance Functor (MyArrow f) where
+    fmap :: (a -> b) -> MyArrow f a -> MyArrow f b
+    fmap f (MyArrow g)  = MyArrow (f . g)
+
+{-
+  1) fmap id = id
+  
+  fmap id (a -> b) = id . a -> b = a -> b
```
Comment (kajigor) [inline Main.hs:43]: what is `a -> b`? This is not a value, it's a type

Comment (kajigor) [inline Main.hs:43]: what is a -> b? This is not a value, it's a type

File: Main.hs [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+
+resolve (Right (Pow (Const n1) (Const n2))) _
+  | isNaN (n1 ** n2)= Left (PowNaN ("(pow of " ++ show n1 ++ " and " ++ show n2 ++ ")"))
+  | otherwise = Right (Const (n1 ** n2))
+
+resolve (Right (Var v)) vars
+  | isNothing lookup = Left (VariableNotApplied ("(variable " ++ v ++ ")"))
+  | otherwise = Right $ Const (snd $ fromJust lookup)
+  where
+    lookup = find (\(x, _) -> x == v) vars
```
Comment (kajigor) [inline Main.hs:118]: There is a function in Prelude which does exactly this.

File: Main.hs [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+                             then Const 1 else Div simple1 simple2
+  where simple1 = simplify expr1
+        simple2 = simplify expr2
+-- same for the Pow
+
+simplify (Root (Const 1)) = Const 1
+simplify (Root (Const 0)) = Const 0
+
+simplify (Root expr) = Root $ simplify expr
+simplify (Plus expr1 expr2) = Plus (simplify expr1) (simplify expr2) -- I wonder if there is a prettier way
```
Comment (kajigor) [inline Main.hs:168]: Using a different representation for the `Expr` data type :)

File: Main.hs [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+    -- (1 - 1^3/3!) * 6
+    (Mult (Const 6) (Minus (Const 1) (Div (Pow (Const 1) (Const 3)) (Mult (Const 2) (Const 3)))), [], Right 5.0),
+    (Root (Const (-1)), [], Left $ NegativeFromRoot "(root of -1.0)"),
+    (Div (Mult (Const 7.0) (Const 2.0)) (Minus (Const 4.0) (Const 4.0)), [], Left $ DivideByZero "(div of 14.0 and 0.0)"),
+    (Div (Pow (Const (-1.0)) (Const 0.5)) (Root (Const 4.0)), [], Left $ PowNaN "(pow of -1.0 and 0.5)"),
+    (Plus (Root (Const (-100500))) (Div (Const 100) (Const 0)), [], Left $ ComplexError (NegativeFromRoot "(root of -100500.0)") (DivideByZero "(div of 100.0 and 0.0)")),
+  
+    (Var "A", [("A", 1)], Right 1),
+    (Var "A", [], Left $ VariableNotApplied "(variable A)"),
+    (Mult (Const 6) (Minus (Var "A") (Div (Pow (Var "A") (Var "C")) (Mult (Var "B") (Var "C")))), [("A", 1), ("B", 2), ("C", 3)], Right 5.0)
```
Comment (kajigor) [inline Main.hs:189]: There is another significant case for variables.

**Review by kajigor — CHANGES_REQUESTED** [review]

**Comment by kajigor** [comment]
Screenshot 2024-03-05 at 11 11 44 AM

There are also at least two issues with simplification. These issues should be fixed and tests for them should be added.

File: Main.hs (hunk 3 above)
Comment (kajigor) [reply Main.hs:168]: (You don't need to rewrite the whole thing now, I'm gonna talk about it tomorrow, actually)

**Comment by kajigor** [comment]
There isn't a solution for the sixth task.

File: Main.hs [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+simplify (Div (Const 0) expr) = if simplify expr /= Const 0 then Const 0 else Div (Const 0) (Const 0)
+
+simplify (Root (Const 1)) = Const 1
+simplify (Root (Const 0)) = Const 0
+
+simplify (Root expr) = Root $ simplify expr
+simplify (Plus expr1 expr2) = Plus (simplify expr1) (simplify expr2) -- I wonder if there is a prettier way
+simplify (Minus expr1 expr2) = Minus (simplify expr1) (simplify expr2)
+simplify (Mult expr1 expr2) = Mult (simplify expr1) (simplify expr2)
+simplify (Div expr1 expr2) = if (simple1 == simple2) && (simple1 /= Const 0) && (simple2 /= Const 0)
```
Comment (kajigor) [inline Main.hs:177]: You only need to check one of the simple's for not equality to 0, btw

**Comment by kajigor** [comment]
ok

**Comment by VenInf** [comment]
Was my "Num" implementation incorrect?

**Pull request closed without merging**